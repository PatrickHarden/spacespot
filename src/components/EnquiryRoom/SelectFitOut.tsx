import React, { useState, useEffect } from 'react'
import { get } from 'lodash'
import { useIntl } from 'react-intl'
import { useSelector, useDispatch } from 'react-redux'
import styled from 'styled-components'
import userSelectors from 'services/user/selectors'
import spaceSelector, { getCountry } from 'services/space/selectors'
import enquirySelectors from 'services/enquiry/selectors'

import negotiationSelectors from 'services/negotiation/selectors'
import actions from 'services/negotiation/actions'
import device from 'services/global/device'

import FitoutForm from '../OnBoarding/FitoutForm'
import PlusIcon from 'components/icons/Plus'
import OutlinedButton from 'components/common/OutlinedButton'
import {
  FitoutOptionProps,
  SelectedFitoutOption,
  TermStatus,
} from 'services/negotiation/types'
import { getRegionCurrencyCode } from 'services/global/region'
import CheckCircle from '@material-ui/icons/CheckCircle'
import Clear from '@material-ui/icons/Clear'
import { formatCurrency } from 'services/global/util'
import HTMLContent from 'components/common/HTMLContent'

const EachFitout = styled.div<{ isLandlord?: boolean }>`
  position: relative;
  box-sizing: border-box;
  border: 1px solid #dddddd;
  border-radius: 2px;
  padding: 20px;
  margin: 10px 0;
  background-color: ${props => (props.isLandlord ? '#FFFFFF' : '#F4F4F4')};
  cursor: ${props => (props.isLandlord ? 'auto' : 'pointer')};
`
const FitoutSelected = styled.div`
  padding: 20px;
  box-sizing: border-box;
  border: 2px solid #6cb9d5;
  border-radius: 2px;
  position: relative;
  margin-bottom: 10px;
`

const IconSelected = styled(CheckCircle)`
  position: absolute;
  background-color: #ffffff;
  top: -11px;
  right: -11px;
  &.MuiSvgIcon-root {
    font-size: 1.5rem;
  }
`
const FitoutTitle = styled.div`
  color: #000000;
  font-size: 18px;
  font-weight: 500;
  line-height: 23px;
`
const Head3 = styled.h3`
  color: #000000;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
`

const FitoutAmount = styled.div`
  color: #000000;
  font-size: 18px;
  font-weight: 500;
  line-height: 23px;
  float: right;
`

const FitoutDescription = styled.div`
  margin-top: 10px;
  color: #6d6c6c;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
`

const ButtonChange = styled.button`
  padding-top: 0;
  color: #6d6c6c;
  border: none;
  outline: none;
  background: white;
  font-size: 16px;
  font-weight: 600;
  text-transform: uppercase;
  cursor: pointer;
`

const SelectedText = styled.div`
  color: #6cb9d5;
  font-weight: 400;
  line-height: 21px;
  float: right;
`

const SkipFloat = styled.div`
  margin-top: 40px;
`

const Container = styled.div`
  @media ${device.mobile} {
    margin: 0;
  }
`

const Plus = styled(PlusIcon)`
  margin: -2px 5px;
`
const ClearIcon = styled(Clear)`
  color: #fff;
  font-size: 15px;
  font-weight: 500;
  position: absolute;
  top: 10px;
  right: 10px;
  cursor: pointer;
  color: #dddddd;
  background-color: trasparent;
`
const TextEstimatedCost = styled.div`
  color: #6d6c6c;
  font-size: 16px;
  font-weight: 400;
  margin-top: 20px;
`
const TextExplanationFitout = styled.div`
  color: #6d6c6c;
  font-size: 16px;
  font-weight: 400;
`
const getSelectedFitout = (
  fitoutOptions: Array<FitoutOptionProps>,
  fitout?: SelectedFitoutOption,
) => {
  if (!fitout || !fitoutOptions) return false
  return fitoutOptions.find(
    (eachFitout: FitoutOptionProps) =>
      Number(eachFitout.id) === fitout.fitoutOption &&
      eachFitout.isCustom === fitout.custom,
  )
}

const SelectFitOut = (props: { disabled: boolean }) => {
  const { disabled } = props
  const dispatch = useDispatch()
  const { formatMessage } = useIntl()
  const t = (s: string) => formatMessage({ id: s })
  const isLandlord = useSelector(userSelectors.isLandlord)
  // FIXME select fitout from enquiry space
  const fitoutOptions = useSelector(spaceSelector.selectSpaceFitout)
  const [showForm, setShowForm] = useState(false)
  const enquiry = useSelector(enquirySelectors.getEnquiry)
  const enquiryStatus = get(enquiry, 'enquireStatus')

  const saveFitout = (fitout: FitoutOptionProps) => {
    setShowForm(false)
    dispatch(actions.putCustomFitout(fitout))
  }
  const selectFitout = (fitout: FitoutOptionProps) => {
    if (!isLandlord && !disabled) {
      dispatch(
        actions.selectFitout({
          custom: fitout.isCustom || false,
          fitoutOption: fitout.id,
        }),
      )
    }
  }
  const unselectFitout = () => {
    if (!isLandlord) {
      dispatch(
        actions.selectFitout({
          custom: false,
          fitoutOption: 0,
        }),
      )
    }
  }
  const deleteFituot = (fitout: FitoutOptionProps) => {
    if (isLandlord && fitout.isCustom) {
      dispatch(actions.deleteFitout(fitout.id as string))
    }
  }

  const space = useSelector(spaceSelector.selectedSpace)
  const country = getCountry(space)
  const currencyCode = getRegionCurrencyCode(country)

  // FIXME this dispatchs load actions every mount
  // use a selector that returns the thing | undefined
  // and dispatch only if undefined
  // or use a custom hook
  const selectedFitout = useSelector(negotiationSelectors.getNegotiationFitout)
  useEffect(() => {
    dispatch(actions.getCustomFitout())
  }, [dispatch])

  const selectedDataFitout = getSelectedFitout(fitoutOptions, selectedFitout)
  const noOptions = !fitoutOptions || fitoutOptions.length === 0

  if (enquiryStatus === TermStatus.Pending) {
    return (
      <Container>
        <p>{t('NEGOTIATE_FIT_OUT_PENDING_MSG')}</p>
      </Container>
    )
  }

  return (
    <Container>
      {selectedDataFitout && (
        <>
          <FitoutSelected>
            {!disabled && (
              <IconSelected fontSize="default" htmlColor="#6CB9D5" />
            )}
            <FitoutAmount>
              {formatCurrency(selectedDataFitout.amount, currencyCode)}
              {!isLandlord ? '*' : ''}
            </FitoutAmount>
            <FitoutTitle>{selectedDataFitout.name}</FitoutTitle>
            <FitoutDescription>
              <HTMLContent html={selectedDataFitout.description} />
            </FitoutDescription>
          </FitoutSelected>
          {!disabled && (
            <>
              {!isLandlord ? (
                <ButtonChange onClick={() => unselectFitout()}>
                  {t('CUSTOM_FITOUT_CHANGE_SELECTION')}
                </ButtonChange>
              ) : null}
              <SelectedText>{t('CUSTOM_FITOUT_SELECTED')}</SelectedText>
            </>
          )}
        </>
      )}
      {(!selectedDataFitout || isLandlord) &&
        fitoutOptions &&
        fitoutOptions.map((fitout: FitoutOptionProps) =>
          !selectedDataFitout ||
          fitout.isCustom !== selectedDataFitout.isCustom ||
          fitout.id !== selectedDataFitout.id ? (
            <EachFitout
              isLandlord={isLandlord}
              key={fitout.name}
              onClick={() => selectFitout(fitout)}>
              {isLandlord && fitout.isCustom ? (
                <ClearIcon onClick={() => deleteFituot(fitout)} />
              ) : null}
              <FitoutAmount>
                {formatCurrency(fitout.amount, currencyCode)}
                {!isLandlord ? '*' : ''}
              </FitoutAmount>
              <FitoutTitle>{fitout.name}</FitoutTitle>
              <FitoutDescription>
                <HTMLContent html={fitout.description} />
              </FitoutDescription>
            </EachFitout>
          ) : null,
        )}
      {!isLandlord && fitoutOptions && fitoutOptions.length >= 1 ? (
        <TextEstimatedCost>{t('NEGOTIATE_ESTIMATED_COST')}</TextEstimatedCost>
      ) : null}
      {noOptions && (
        <>
          {isLandlord ? (
            <Head3>{t('CUSTOM_FITOUT_EMPTY')}</Head3>
          ) : (
            <TextExplanationFitout>
              {t('NEGOTIATE_EXPLANATION_FITOUT')}
            </TextExplanationFitout>
          )}
        </>
      )}
      {!disabled && isLandlord && (
        <>
          {showForm ? (
            <SkipFloat>
              <FitoutForm
                setShow={() => setShowForm(false)}
                saveFitout={saveFitout}
              />
            </SkipFloat>
          ) : (
            <OutlinedButton onClick={() => setShowForm(true)}>
              <Plus size="1.1em" /> {t('CUSTOM_FITOUT_OPTION_ADD')}
            </OutlinedButton>
          )}
        </>
      )}
    </Container>
  )
}

export default SelectFitOut
