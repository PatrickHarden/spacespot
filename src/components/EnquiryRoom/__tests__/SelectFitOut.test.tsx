/* eslint-disable @typescript-eslint/camelcase */
import React from 'react'
import { mount } from 'enzyme'
import { mountToJson } from 'enzyme-to-json'

import SelectFitOut from '../SelectFitOut'

jest.mock('react-intl', () => ({
  ...jest.requireActual('react-intl'),
  useIntl: () => ({ formatMessage: (s: any) => s.id }),
}))
let mockState = {
  user: { auth: { idToken: { claims: { extension_Landlord: false } } } },
  space: {},
}

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => jest.fn(),
  useSelector: (selector: any) => selector(mockState),
}))
it('renders without crashing tenant', async () => {
  const wrapper = mount(<SelectFitOut />)
  expect(mountToJson(wrapper)).toMatchSnapshot()
})
it('renders without crashing Landlord', async () => {
  mockState.user.auth.idToken.claims.extension_Landlord = true
  const wrapper = mount(<SelectFitOut />)
  expect(mountToJson(wrapper)).toMatchSnapshot()
})
it('renders without crashing tenant with common fitout', async () => {
  mockState = {
    user: { auth: { idToken: { claims: { extension_Landlord: false } } } },
    space: {
      item: {
        Document: {
          'Common.FitOutOptions': [
            {
              'Common.Identifier': 1,
              'Common.Name': [{ 'Common.Text': 'a' }],
              'Common.Description': [{ 'Common.Text': 'a' }],
              'Common.Charges': [{ 'Common.Amount': 'a' }],
            },
          ],
        },
      },
    },
  }

  const wrapper = mount(<SelectFitOut />)
  expect(mountToJson(wrapper)).toMatchSnapshot()
})
