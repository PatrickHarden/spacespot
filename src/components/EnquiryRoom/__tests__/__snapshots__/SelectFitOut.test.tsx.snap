// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing Landlord 1`] = `
<SelectFitOut>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-hMqMXs",
            "isStatic": false,
            "lastClassName": "caeSfG",
            "rules": Array [
              "
  margin-left: 20px;
  @media ",
              "(max-width: 767px)",
              " {
    margin: 0;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-hMqMXs",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-hMqMXs caeSfG"
      >
        <styled.h3>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-kGXeez",
                  "isStatic": false,
                  "lastClassName": "IZhOg",
                  "rules": Array [
                    "
  color: #000000;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
",
                  ],
                },
                "displayName": "styled.h3",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-kGXeez",
                "target": "h3",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <h3
              className="sc-kGXeez IZhOg"
            >
              CUSTOM_FITOUT_EMPTY
            </h3>
          </StyledComponent>
        </styled.h3>
        <styled.button
          onClick={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-jzJRlG",
                  "isStatic": false,
                  "lastClassName": "eaXYXz",
                  "rules": Array [
                    "
  font-size: 14px;
  font-weight: 550;
  border: 1px solid #6cb9d5;
  color: #6cb9d5;
  background-color: #ffffff;
  border-radius: 2px;
  cursor: pointer;
  text-transform: uppercase;
  height: 37px;
  padding: 0 20px;
",
                  ],
                },
                "displayName": "styled.button",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-jzJRlG",
                "target": "button",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onClick={[Function]}
          >
            <button
              className="sc-jzJRlG eaXYXz"
              onClick={[Function]}
            >
              <Styled(Plus)
                size="1.1em"
              >
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kEYyzF",
                        "isStatic": false,
                        "lastClassName": "deUVHB",
                        "rules": Array [
                          "
  margin: -2px 5px;
",
                        ],
                      },
                      "displayName": "Styled(Plus)",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kEYyzF",
                      "target": [Function],
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  size="1.1em"
                >
                  <Plus
                    className="sc-kEYyzF deUVHB"
                    size="1.1em"
                  >
                    <svg
                      className="sc-kEYyzF deUVHB"
                      height="1.1em"
                      size="1.1em"
                      viewBox="0 0 18 17"
                      width="1.1em"
                    >
                      <g
                        fill="none"
                        fillRule="evenodd"
                        stroke="#6CB9D5"
                        strokeLinecap="round"
                        strokeWidth={2}
                      >
                        <path
                          d="M9 1.6v13.8M1.64 8.5h14.72"
                        />
                      </g>
                    </svg>
                  </Plus>
                </StyledComponent>
              </Styled(Plus)>
               
              CUSTOM_FITOUT_OPTION_ADD
            </button>
          </StyledComponent>
        </styled.button>
      </div>
    </StyledComponent>
  </styled.div>
</SelectFitOut>
`;

exports[`renders without crashing tenant 1`] = `
<SelectFitOut>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-hMqMXs",
            "isStatic": false,
            "lastClassName": "caeSfG",
            "rules": Array [
              "
  margin-left: 20px;
  @media ",
              "(max-width: 767px)",
              " {
    margin: 0;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-hMqMXs",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-hMqMXs caeSfG"
      >
        <styled.div>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-hSdWYo",
                  "isStatic": false,
                  "lastClassName": "jRRjJN",
                  "rules": Array [
                    "
  color: #6d6c6c;
  font-size: 16px;
  font-weight: 400;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-hSdWYo",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <div
              className="sc-hSdWYo jRRjJN"
            >
              NEGOTIATE_EXPLANATION_FITOUT
            </div>
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</SelectFitOut>
`;

exports[`renders without crashing tenant with common fitout 1`] = `
<SelectFitOut>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-hMqMXs",
            "isStatic": false,
            "lastClassName": "caeSfG",
            "rules": Array [
              "
  margin-left: 20px;
  @media ",
              "(max-width: 767px)",
              " {
    margin: 0;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-hMqMXs",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-hMqMXs caeSfG"
      >
        <styled.div
          isLandlord={false}
          key="a"
          onClick={[Function]}
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-cSHVUG",
                  "isStatic": false,
                  "lastClassName": "blQaBk",
                  "rules": Array [
                    "
  position: relative;
  box-sizing: border-box;
  border: 1px solid #dddddd;
  border-radius: 2px;
  padding: 20px;
  margin: 10px 0;
  background-color: ",
                    [Function],
                    ";
  cursor: ",
                    [Function],
                    ";
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-cSHVUG",
                "target": "div",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            isLandlord={false}
            onClick={[Function]}
          >
            <div
              className="sc-cSHVUG blQaBk"
              onClick={[Function]}
            >
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kpOJdX",
                        "isStatic": false,
                        "lastClassName": "bkkRbR",
                        "rules": Array [
                          "
  color: #000000;
  font-size: 18px;
  font-weight: 500;
  line-height: 23px;
  float: right;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kpOJdX",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-kpOJdX bkkRbR"
                  >
                    akr
                    *
                  </div>
                </StyledComponent>
              </styled.div>
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-kgoBCf",
                        "isStatic": false,
                        "lastClassName": "dZSsVL",
                        "rules": Array [
                          "
  color: #000000;
  font-size: 18px;
  font-weight: 500;
  line-height: 23px;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-kgoBCf",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-kgoBCf dZSsVL"
                  >
                    a
                  </div>
                </StyledComponent>
              </styled.div>
              <styled.div>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-dxgOiQ",
                        "isStatic": false,
                        "lastClassName": "gfnofS",
                        "rules": Array [
                          "
  margin-top: 10px;
  color: #6d6c6c;
  font-size: 16px;
  font-weight: 400;
  line-height: 21px;
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-dxgOiQ",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <div
                    className="sc-dxgOiQ gfnofS"
                  >
                    <HTMLContent
                      html="a"
                    >
                      <div
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "a",
                          }
                        }
                      />
                    </HTMLContent>
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</SelectFitOut>
`;
